File: jetbrains-plugin/src/main/kotlin/com/vscode/jetbrainssync/IDEPathDetector.kt
package com.vscode.jetbrainssync

import com.intellij.openapi.diagnostic.Logger
import java.io.File

/**
 * IDE Path Detector
 * Automatically detects installation paths for VSCode/Cursor/Windsurf
 */
class IDEPathDetector {
    private val log: Logger = Logger.getInstance(IDEPathDetector::class.java)
    private val osName: String = System.getProperty("os.name").lowercase()
    private val isWindows: Boolean = osName.contains("windows")
    private val isMacOS: Boolean = osName.contains("mac")
    private val isLinux: Boolean = osName.contains("linux") || osName.contains("unix")
    private val userHome: String = System.getProperty("user.home")

    /**
     * Detect VSCode path
     * Try different IDEs by priority (VSCode, Cursor, Windsurf)
     */
    fun detectVSCodePath(): String? {
        log.info("Starting to detect VSCode/Cursor/Windsurf paths...")

        val detectedPath = findFirstAvailableIDE()

        if (detectedPath != null) {
            log.info("Detected IDE: $detectedPath")
        } else {
            log.warn("No VSCode/Cursor/Windsurf detected")
        }

        return detectedPath
    }

    /**
     * Find first available IDE
     */
    private fun findFirstAvailableIDE(): String? {
        val candidates = getIDECandidates()

        for (candidate in candidates) {
            val expandedPaths = expandPath(candidate)
            for (path in expandedPaths) {
                if (isExecutable(path)) {
                    return path
                }
            }
        }

        return null
    }

    /**
     * Get IDE candidate path list
     */
    private fun getIDECandidates(): List<String> {
        return when {
            isMacOS -> getMacOSCandidates()
            isWindows -> getWindowsCandidates()
            isLinux -> getLinuxCandidates()
            else -> emptyList()
        }
    }

    /**
     * macOS IDE candidate paths
     */
    private fun getMacOSCandidates(): List<String> {
        return listOf(
            // Standard installation path - Visual Studio Code
            "/Applications/Visual Studio Code.app/Contents/Resources/app/bin/code",
            "/Applications/VSCode.app/Contents/Resources/app/bin/code",
            
            // Standard installation path - Cursor
            "/Applications/Cursor.app/Contents/Resources/app/bin/cursor",
            
            // Standard installation path - Windsurf
            "/Applications/Windsurf.app/Contents/Resources/app/bin/windsurf",
            
            // User installation paths
            "$userHome/Applications/Visual Studio Code.app/Contents/Resources/app/bin/code",
            "$userHome/Applications/Cursor.app/Contents/Resources/app/bin/cursor",
            "$userHome/Applications/Windsurf.app/Contents/Resources/app/bin/windsurf",
            
            // CLI command paths
            "/usr/local/bin/code",
            "/usr/local/bin/cursor",
            "/usr/local/bin/windsurf",
            
            // Homebrew paths
            "/opt/homebrew/bin/code",
            "/opt/homebrew/bin/cursor",
            "/opt/homebrew/bin/windsurf"
        )
    }

    /**
     * Windows IDE candidate paths
     */
    private fun getWindowsCandidates(): List<String> {
        val localAppData = System.getenv("LOCALAPPDATA") ?: "$userHome\\AppData\\Local"
        val programFiles = System.getenv("ProgramFiles") ?: "C:\\Program Files"
        val programFilesX86 = System.getenv("ProgramFiles(x86)") ?: "C:\\Program Files (x86)"

        return listOf(
            // Standard installation path - Visual Studio Code
            "$localAppData\\Programs\\Microsoft VS Code\\bin\\code.cmd",
            "$localAppData\\Programs\\Microsoft VS Code\\Code.exe",
            "$programFiles\\Microsoft VS Code\\bin\\code.cmd",
            "$programFiles\\Microsoft VS Code\\Code.exe",
            
            // Standard installation path - Cursor
            "$localAppData\\Programs\\cursor\\Cursor.exe",
            "$localAppData\\cursor\\Cursor.exe",
            
            // Standard installation path - Windsurf
            "$localAppData\\Programs\\Windsurf\\Windsurf.exe",
            "$localAppData\\Windsurf\\Windsurf.exe",
            
            // System PATH command paths
            "code.cmd",
            "cursor.cmd",
            "windsurf.cmd"
        )
    }

    /**
     * Linux IDE candidate paths
     */
    private fun getLinuxCandidates(): List<String> {
        return listOf(
            // System binary paths
            "/usr/bin/code",
            "/usr/bin/cursor",
            "/usr/bin/windsurf",
            
            // Local binary paths
            "/usr/local/bin/code",
            "/usr/local/bin/cursor",
            "/usr/local/bin/windsurf",
            
            // Snap installations
            "/snap/code/current/bin/code",
            "/snap/bin/code",
            
            // Flatpak installations
            "/var/lib/flatpak/exports/bin/com.visualstudio.code",
            
            // User local installations
            "$userHome/.local/bin/code",
            "$userHome/.local/bin/cursor",
            "$userHome/.local/bin/windsurf",
            
            // Manual installation paths
            "/opt/VSCode-linux-x64/bin/code",
            "/opt/cursor/cursor",
            "/opt/windsurf/windsurf"
        )
    }

    /**
     * Expand paths containing wildcards
     */
    private fun expandPath(pattern: String): List<String> {
        if (!pattern.contains("*")) {
            return listOf(pattern)
        }

        val results = mutableListOf<String>()
        val parts = pattern.split(File.separator)
        var currentPaths = mutableListOf("")

        for (part in parts) {
            if (part.isEmpty()) {
                currentPaths = mutableListOf(File.separator)
                continue
            }

            val newPaths = mutableListOf<String>()

            for (currentPath in currentPaths) {
                if (part.contains("*")) {
                    // Expand wildcard
                    val dir = if (currentPath.isEmpty()) "." else currentPath
                    val directory = File(dir)
                    
                    if (directory.exists() && directory.isDirectory) {
                        val regex = Regex("^" + part.replace("*", ".*") + "$")
                        val entries = directory.listFiles() ?: emptyArray()
                        
                        for (entry in entries) {
                            if (regex.matches(entry.name)) {
                                newPaths.add(File(currentPath, entry.name).path)
                            }
                        }
                    }
                } else {
                    newPaths.add(File(currentPath, part).path)
                }
            }

            currentPaths = newPaths
        }

        return currentPaths
    }

    /**
     * Check if file is executable
     */
    private fun isExecutable(path: String): Boolean {
        val file = File(path)
        return file.exists() && file.canExecute()
    }

    /**
     * Detect JetBrains IDE path (for VSCode use)
     */
    fun detectJetBrainsPath(): String? {
        val candidates = getJetBrainsCandidates()

        for (candidate in candidates) {
            val expandedPaths = expandPath(candidate)
            for (path in expandedPaths) {
                if (isExecutable(path)) {
                    return path
                }
            }
        }

        return null
    }

    /**
     * JetBrains IDE candidate paths
     */
    private fun getJetBrainsCandidates(): List<String> {
        return when {
            isMacOS -> listOf(
                "$userHome/Library/Application Support/JetBrains/Toolbox/scripts/idea",
                "$userHome/Library/Application Support/JetBrains/Toolbox/scripts/webstorm",
                "$userHome/Library/Application Support/JetBrains/Toolbox/scripts/pycharm",
                "/Applications/IntelliJ IDEA.app/Contents/MacOS/idea",
                "/Applications/WebStorm.app/Contents/MacOS/webstorm",
                "/Applications/PyCharm.app/Contents/MacOS/pycharm"
            )
            isWindows -> {
                val localAppData = System.getenv("LOCALAPPDATA") ?: "$userHome\\AppData\\Local"
                listOf(
                    "$localAppData\\JetBrains\\Toolbox\\scripts\\idea.cmd",
                    "$localAppData\\JetBrains\\Toolbox\\scripts\\webstorm.cmd",
                    "$localAppData\\JetBrains\\Toolbox\\scripts\\pycharm.cmd"
                )
            }
            isLinux -> listOf(
                "$userHome/.local/share/JetBrains/Toolbox/scripts/idea",
                "$userHome/.local/share/JetBrains/Toolbox/scripts/webstorm",
                "$userHome/.local/share/JetBrains/Toolbox/scripts/pycharm",
                "/snap/intellij-idea-ultimate/current/bin/idea.sh",
                "/snap/intellij-idea-community/current/bin/idea.sh"
            )
            else -> emptyList()
        }
    }

    /**
     * Data class representing a detected IDE
     */
    data class DetectedIDE(
        val name: String,
        val path: String
    )

    /**
     * Detect all available VSCode/Cursor/Windsurf IDEs
     * Returns a list of detected IDEs with names and paths
     */
    fun detectAllVSCodePaths(): List<DetectedIDE> {
        log.info("Detecting all VSCode/Cursor/Windsurf IDEs...")
        val results = mutableListOf<DetectedIDE>()
        val candidates = getIDECandidates()

        for (candidate in candidates) {
            val expandedPaths = expandPath(candidate)
            for (path in expandedPaths) {
                if (isExecutable(path)) {
                    val name = getIDENameFromPath(path)
                    // Avoid duplicates by name
                    if (results.none { it.name == name }) {
                        results.add(DetectedIDE(name, path))
                        log.info("Detected IDE: $name at $path")
                    }
                }
            }
        }

        return results
    }

    /**
     * Get IDE name from path
     */
    private fun getIDENameFromPath(path: String): String {
        val lowerPath = path.lowercase()
        return when {
            lowerPath.contains("windsurf") -> "Windsurf"
            lowerPath.contains("cursor") -> "Cursor"
            lowerPath.contains("visual studio code") || lowerPath.contains("vs code") -> "Visual Studio Code"
            lowerPath.endsWith("/code") || lowerPath.endsWith("\\code") || 
            lowerPath.endsWith("/code.cmd") || lowerPath.endsWith("\\code.cmd") ||
            lowerPath.endsWith("code.exe") -> "Visual Studio Code"
            else -> File(path).nameWithoutExtension.replaceFirstChar { it.uppercase() }
        }
    }

    /**
     * Resolve a command name or path to an executable path
     * Supports:
     * - Full paths (e.g., "/usr/bin/windsurf")
     * - Command names (e.g., "windsurf", "cursor", "code")
     */
    fun resolveIDEPath(pathOrCommand: String): String? {
        if (pathOrCommand.isBlank()) {
            return null
        }

        // First, check if it's already a valid executable path
        if (File(pathOrCommand).let { it.exists() && it.canExecute() }) {
            log.info("Path is already executable: $pathOrCommand")
            return pathOrCommand
        }

        // Check if it looks like a command name (no path separators)
        if (!pathOrCommand.contains(File.separator) && !pathOrCommand.contains("/") && !pathOrCommand.contains("\\")) {
            log.info("Attempting to resolve command name: $pathOrCommand")
            
            // Try to find in PATH using which/where
            val resolvedPath = findCommandInPath(pathOrCommand)
            if (resolvedPath != null) {
                log.info("Resolved command '$pathOrCommand' to: $resolvedPath")
                return resolvedPath
            }

            // Try to find in known IDE paths
            val knownPath = findInKnownPaths(pathOrCommand)
            if (knownPath != null) {
                log.info("Found '$pathOrCommand' in known paths: $knownPath")
                return knownPath
            }
        }

        log.warn("Could not resolve IDE path: $pathOrCommand")
        return null
    }

    /**
     * Find command in system PATH using which (Unix) or where (Windows)
     */
    private fun findCommandInPath(command: String): String? {
        return try {
            val cmd = if (isWindows) arrayOf("cmd", "/c", "where", command) else arrayOf("which", command)
            val process = ProcessBuilder(*cmd)
                .redirectErrorStream(true)
                .start()
            
            val result = process.inputStream.bufferedReader().readLine()
            val exitCode = process.waitFor()
            
            if (exitCode == 0 && !result.isNullOrBlank() && File(result).exists()) {
                result
            } else {
                null
            }
        } catch (e: Exception) {
            log.debug("Failed to execute which/where for '$command': ${e.message}")
            null
        }
    }

    /**
     * Find command in known IDE installation paths
     */
    private fun findInKnownPaths(command: String): String? {
        val lowerCommand = command.lowercase()
        val candidates = getIDECandidates()

        for (candidate in candidates) {
            val expandedPaths = expandPath(candidate)
            for (path in expandedPaths) {
                val lowerPath = path.lowercase()
                // Match by command name in the path
                if (lowerPath.contains(lowerCommand) && isExecutable(path)) {
                    return path
                }
            }
        }

        return null
    }
}

File: vscode-extension/src/IDEPathDetector.ts
import * as fs from 'fs';
import * as path from 'path';
import * as os from 'os';
import {Logger} from './Logger';

/**
 * IDE Path Detector
 * Automatically detects installation paths for JetBrains IDEs
 */
export class IDEPathDetector {
    private readonly logger: Logger;
    private readonly platform: NodeJS.Platform;

    constructor(logger: Logger) {
        this.logger = logger;
        this.platform = process.platform;
    }

    /**
     * Detect JetBrains IDE path
     * Try different IDEs by priority (IntelliJ IDEA, WebStorm, PyCharm, Rider, GoLand, PhpStorm, CLion, RubyMine)
     */
    async detectJetBrainsPath(): Promise<string | null> {
        this.logger.info('Detecting JetBrains IDE path...');

        const detectedPath = await this.findFirstAvailableIDE();
        
        if (detectedPath) {
            this.logger.info(`Detected JetBrains IDE: ${detectedPath}`);
        } else {
            this.logger.warn('No JetBrains IDE detected');
        }

        return detectedPath;
    }

    /**
     * Find first available IDE
     */
    private async findFirstAvailableIDE(): Promise<string | null> {
        const candidates = this.getIDECandidates();

        for (const candidate of candidates) {
            const expandedPaths = await this.expandPath(candidate);
            for (const expandedPath of expandedPaths) {
                if (await this.fileExists(expandedPath)) {
                    return expandedPath;
                }
            }
        }

        return null;
    }

    /**
     * Get IDE candidate path list
     */
    private getIDECandidates(): string[] {
        switch (this.platform) {
            case 'darwin':
                return this.getMacOSCandidates();
            case 'win32':
                return this.getWindowsCandidates();
            case 'linux':
                return this.getLinuxCandidates();
            default:
                return [];
        }
    }

    /**
     * macOS IDE candidate paths
     */
    private getMacOSCandidates(): string[] {
        const home = os.homedir();
        return [
            // JetBrains Toolbox installation paths
            `${home}/Library/Application Support/JetBrains/Toolbox/scripts/idea`,
            `${home}/Library/Application Support/JetBrains/Toolbox/scripts/webstorm`,
            `${home}/Library/Application Support/JetBrains/Toolbox/scripts/pycharm`,
            `${home}/Library/Application Support/JetBrains/Toolbox/scripts/rider`,
            `${home}/Library/Application Support/JetBrains/Toolbox/scripts/goland`,
            `${home}/Library/Application Support/JetBrains/Toolbox/scripts/phpstorm`,
            `${home}/Library/Application Support/JetBrains/Toolbox/scripts/clion`,
            `${home}/Library/Application Support/JetBrains/Toolbox/scripts/rubymine`,
            
            // Standard installation paths - IntelliJ IDEA
            '/Applications/IntelliJ IDEA.app/Contents/MacOS/idea',
            '/Applications/IntelliJ IDEA CE.app/Contents/MacOS/idea',
            '/Applications/IntelliJ IDEA Ultimate.app/Contents/MacOS/idea',
            
            // Standard installation paths - WebStorm
            '/Applications/WebStorm.app/Contents/MacOS/webstorm',
            
            // Standard installation paths - PyCharm
            '/Applications/PyCharm.app/Contents/MacOS/pycharm',
            '/Applications/PyCharm CE.app/Contents/MacOS/pycharm',
            '/Applications/PyCharm Professional.app/Contents/MacOS/pycharm',
            
            // Standard installation paths - Rider
            '/Applications/Rider.app/Contents/MacOS/rider',
            
            // Standard installation paths - GoLand
            '/Applications/GoLand.app/Contents/MacOS/goland',
            
            // Standard installation paths - PhpStorm
            '/Applications/PhpStorm.app/Contents/MacOS/phpstorm',
            
            // Standard installation paths - CLion
            '/Applications/CLion.app/Contents/MacOS/clion',
            
            // Standard installation paths - RubyMine
            '/Applications/RubyMine.app/Contents/MacOS/rubymine',
            
            // User installation paths
            `${home}/Applications/IntelliJ IDEA.app/Contents/MacOS/idea`,
            `${home}/Applications/WebStorm.app/Contents/MacOS/webstorm`,
            `${home}/Applications/PyCharm.app/Contents/MacOS/pycharm`,
            `${home}/Applications/Rider.app/Contents/MacOS/rider`,
        ];
    }

    /**
     * Windows IDE candidate paths
     */
    private getWindowsCandidates(): string[] {
        const home = os.homedir();
        const programFiles = process.env['ProgramFiles'] || 'C:\\Program Files';
        const programFilesX86 = process.env['ProgramFiles(x86)'] || 'C:\\Program Files (x86)';
        const localAppData = process.env['LOCALAPPDATA'] || `${home}\\AppData\\Local`;

        return [
            // JetBrains Toolbox
            `${localAppData}\\JetBrains\\Toolbox\\scripts\\idea.cmd`,
            `${localAppData}\\JetBrains\\Toolbox\\scripts\\webstorm.cmd`,
            `${localAppData}\\JetBrains\\Toolbox\\scripts\\pycharm.cmd`,
            `${localAppData}\\JetBrains\\Toolbox\\scripts\\rider.cmd`,
            `${localAppData}\\JetBrains\\Toolbox\\scripts\\goland.cmd`,
            `${localAppData}\\JetBrains\\Toolbox\\scripts\\phpstorm.cmd`,
            
            // Standard installation paths (using wildcard patterns)
            `${programFiles}\\JetBrains\\IntelliJ IDEA*\\bin\\idea64.exe`,
            `${programFiles}\\JetBrains\\WebStorm*\\bin\\webstorm64.exe`,
            `${programFiles}\\JetBrains\\PyCharm*\\bin\\pycharm64.exe`,
            `${programFiles}\\JetBrains\\Rider*\\bin\\rider64.exe`,
            `${programFiles}\\JetBrains\\GoLand*\\bin\\goland64.exe`,
            `${programFiles}\\JetBrains\\PhpStorm*\\bin\\phpstorm64.exe`,
            `${programFiles}\\JetBrains\\CLion*\\bin\\clion64.exe`,
            
            // x86 paths
            `${programFilesX86}\\JetBrains\\IntelliJ IDEA*\\bin\\idea64.exe`,
            `${programFilesX86}\\JetBrains\\WebStorm*\\bin\\webstorm64.exe`,
        ];
    }

    /**
     * Linux IDE candidate paths
     */
    private getLinuxCandidates(): string[] {
        const home = os.homedir();

        return [
            // JetBrains Toolbox
            `${home}/.local/share/JetBrains/Toolbox/scripts/idea`,
            `${home}/.local/share/JetBrains/Toolbox/scripts/webstorm`,
            `${home}/.local/share/JetBrains/Toolbox/scripts/pycharm`,
            `${home}/.local/share/JetBrains/Toolbox/scripts/rider`,
            `${home}/.local/share/JetBrains/Toolbox/scripts/goland`,
            `${home}/.local/share/JetBrains/Toolbox/scripts/phpstorm`,
            
            // Snap installations
            '/snap/intellij-idea-ultimate/current/bin/idea.sh',
            '/snap/intellij-idea-community/current/bin/idea.sh',
            '/snap/webstorm/current/bin/webstorm.sh',
            '/snap/pycharm-professional/current/bin/pycharm.sh',
            '/snap/pycharm-community/current/bin/pycharm.sh',
            
            // Manual installation paths
            `/opt/idea/bin/idea.sh`,
            `/opt/idea-*/bin/idea.sh`,
            `/opt/webstorm/bin/webstorm.sh`,
            `/opt/pycharm/bin/pycharm.sh`,
            `${home}/idea/bin/idea.sh`,
            `${home}/.local/idea/bin/idea.sh`,
            
            // Standard binary paths
            '/usr/bin/idea',
            '/usr/bin/webstorm',
            '/usr/bin/pycharm',
            '/usr/local/bin/idea',
            '/usr/local/bin/webstorm',
            '/usr/local/bin/pycharm',
        ];
    }

    /**
     * Expand paths containing wildcards
     */
    private async expandPath(pattern: string): Promise<string[]> {
        if (!pattern.includes('*')) {
            return [pattern];
        }

        const results: string[] = [];
        const parts = pattern.split(path.sep);
        let currentPaths = [''];

        for (const part of parts) {
            if (part === '') {
                currentPaths = [path.sep];
                continue;
            }

            const newPaths: string[] = [];

            for (const currentPath of currentPaths) {
                if (part.includes('*')) {
                    // Expand wildcard
                    const dir = currentPath || '.';
                    try {
                        const entries = await fs.promises.readdir(dir);
                        const regex = new RegExp('^' + part.replace(/\*/g, '.*') + '$');
                        
                        for (const entry of entries) {
                            if (regex.test(entry)) {
                                newPaths.push(path.join(currentPath, entry));
                            }
                        }
                    } catch {
                        // Directory doesn't exist, skip
                    }
                } else {
                    newPaths.push(path.join(currentPath, part));
                }
            }

            currentPaths = newPaths;
        }

        return currentPaths;
    }

    /**
     * Check if file is executable
     */
    private async fileExists(filePath: string): Promise<boolean> {
        try {
            await fs.promises.access(filePath, fs.constants.X_OK);
            return true;
        } catch {
            return false;
        }
    }

    /**
     * Detect VSCode path (for JetBrains use)
     */
    async detectVSCodePath(): Promise<string | null> {
        const candidates = this.getVSCodeCandidates();

        for (const candidate of candidates) {
            const expandedPaths = await this.expandPath(candidate);
            for (const expandedPath of expandedPaths) {
                if (await this.fileExists(expandedPath)) {
                    return expandedPath;
                }
            }
        }

        return null;
    }

    /**
     * Get VSCode/Cursor/Windsurf candidate paths
     */
    private getVSCodeCandidates(): string[] {
        const home = os.homedir();

        switch (this.platform) {
            case 'darwin':
                return [
                    '/Applications/Visual Studio Code.app/Contents/Resources/app/bin/code',
                    '/Applications/Cursor.app/Contents/Resources/app/bin/cursor',
                    '/Applications/Windsurf.app/Contents/Resources/app/bin/windsurf',
                    `${home}/Applications/Visual Studio Code.app/Contents/Resources/app/bin/code`,
                    `${home}/Applications/Cursor.app/Contents/Resources/app/bin/cursor`,
                    `${home}/Applications/Windsurf.app/Contents/Resources/app/bin/windsurf`,
                    // CLI commands
                    '/usr/local/bin/code',
                    '/usr/local/bin/cursor',
                    '/usr/local/bin/windsurf',
                ];

            case 'win32':
                const localAppData = process.env['LOCALAPPDATA'] || `${home}\\AppData\\Local`;
                return [
                    `${localAppData}\\Programs\\Microsoft VS Code\\bin\\code.cmd`,
                    `${localAppData}\\Programs\\cursor\\Cursor.exe`,
                    `${localAppData}\\Programs\\Windsurf\\Windsurf.exe`,
                    'C:\\Program Files\\Microsoft VS Code\\bin\\code.cmd',
                ];

            case 'linux':
                return [
                    '/usr/bin/code',
                    '/usr/bin/cursor',
                    '/usr/bin/windsurf',
                    '/snap/code/current/bin/code',
                    `${home}/.local/bin/code`,
                ];

            default:
                return [];
        }
    }

    /**
     * Detect all available JetBrains IDEs
     * Returns a list of detected IDEs with names and paths
     */
    async detectAllJetBrainsPaths(): Promise<DetectedIDE[]> {
        this.logger.info('Detecting all JetBrains IDEs...');
        const results: DetectedIDE[] = [];
        const candidates = this.getIDECandidates();

        for (const candidate of candidates) {
            const expandedPaths = await this.expandPath(candidate);
            for (const expandedPath of expandedPaths) {
                if (await this.fileExists(expandedPath)) {
                    const name = this.getIDENameFromPath(expandedPath);
                    // Avoid duplicates by name
                    if (!results.find(r => r.name === name)) {
                        results.push({ name, path: expandedPath });
                        this.logger.info(`Detected IDE: ${name} at ${expandedPath}`);
                    }
                }
            }
        }

        return results;
    }

    /**
     * Get IDE name from path
     */
    private getIDENameFromPath(idePath: string): string {
        const lowerPath = idePath.toLowerCase();
        
        if (lowerPath.includes('rider')) return 'Rider';
        if (lowerPath.includes('webstorm')) return 'WebStorm';
        if (lowerPath.includes('pycharm')) return 'PyCharm';
        if (lowerPath.includes('goland')) return 'GoLand';
        if (lowerPath.includes('phpstorm')) return 'PhpStorm';
        if (lowerPath.includes('clion')) return 'CLion';
        if (lowerPath.includes('rubymine')) return 'RubyMine';
        if (lowerPath.includes('idea')) return 'IntelliJ IDEA';
        
        // Extract filename
        const basename = path.basename(idePath);
        const name = basename.replace(/\.(exe|cmd|sh|app)$/i, '');
        return name.charAt(0).toUpperCase() + name.slice(1);
    }

    /**
     * Resolve a command name or path to an executable path
     * Supports:
     * - Full paths (e.g., "/opt/idea/bin/idea.sh")
     * - Command names (e.g., "idea", "webstorm", "pycharm")
     */
    async resolveIDEPath(pathOrCommand: string): Promise<string | null> {
        if (!pathOrCommand || pathOrCommand.trim() === '') {
            return null;
        }

        // First, check if it's already a valid executable path
        if (await this.fileExists(pathOrCommand)) {
            this.logger.info(`Path is already executable: ${pathOrCommand}`);
            return pathOrCommand;
        }

        // Check if it looks like a command name (no path separators)
        if (!pathOrCommand.includes(path.sep) && !pathOrCommand.includes('/') && !pathOrCommand.includes('\\')) {
            this.logger.info(`Attempting to resolve command name: ${pathOrCommand}`);
            
            // Try to find in PATH using which/where
            const resolvedPath = await this.findCommandInPath(pathOrCommand);
            if (resolvedPath) {
                this.logger.info(`Resolved command '${pathOrCommand}' to: ${resolvedPath}`);
                return resolvedPath;
            }

            // Try to find in known IDE paths
            const knownPath = await this.findInKnownPaths(pathOrCommand);
            if (knownPath) {
                this.logger.info(`Found '${pathOrCommand}' in known paths: ${knownPath}`);
                return knownPath;
            }
        }

        this.logger.warn(`Could not resolve IDE path: ${pathOrCommand}`);
        return null;
    }

    /**
     * Find command in system PATH using which (Unix) or where (Windows)
     */
    private async findCommandInPath(command: string): Promise<string | null> {
        const { exec } = require('child_process');
        const util = require('util');
        const execAsync = util.promisify(exec);

        try {
            const cmd = this.platform === 'win32' ? `where ${command}` : `which ${command}`;
            const { stdout } = await execAsync(cmd);
            const result = stdout.trim().split('\n')[0];
            
            if (result && await this.fileExists(result)) {
                return result;
            }
        } catch {
            // Command not found in PATH
        }

        return null;
    }

    /**
     * Find command in known IDE installation paths
     */
    private async findInKnownPaths(command: string): Promise<string | null> {
        const lowerCommand = command.toLowerCase();
        const candidates = this.getIDECandidates();

        for (const candidate of candidates) {
            const expandedPaths = await this.expandPath(candidate);
            for (const expandedPath of expandedPaths) {
                const lowerPath = expandedPath.toLowerCase();
                // Match by command name in the path
                if (lowerPath.includes(lowerCommand) && await this.fileExists(expandedPath)) {
                    return expandedPath;
                }
            }
        }

        return null;
    }
}

/**
 * Detected IDE information export
 */
export interface DetectedIDE {
    name: string;
    path: string;
}
